### ISSUE TITLE
Choosing GDScript Formatting and Linter with "Running "Format Document With..." in VSCode Causes Error

### ISSUE THREAD

**Issue by kaiharuto** [EddieDover/vscode-formatter-godot#11]
When trying to open the VSC command shown in the title, VSCode only gives me an error message saying "An unknown error has occured"
Consulting the log files, I see the following

```
setting an array with 0 elements as diag for URI: file:///home/kaiharuto/projects/NESasssin/src/Actor/ActorClass.gd
stderr: /bin/sh: 1: gdformat: not found

err: Error: Command failed: gdformat --line-length=100  -
/bin/sh: 1: gdformat: not found
```
both `gdlint` and `gdformat` work on the command line and I tried appending

`/home/kaiharuto/.local/bin/gdparse: /home/kaiharuto/.local/bin/gdformat: /home/kaiharuto/.local/bin/gdlint:/home/kaiharuto/.local/lib/python3.10/site-packages/gdtoolkit`

to my path as simply having `:/home/kaiharuto/.local/lib/python3.10/site-packages/gdtoolkit:` didn't do the trick either

Has anyone else encountered a similar issue?

**Comment by EddieDover** [comment]
I haven't had any reports of this behavior, but I'll see what I can do about replicating it this week on one of my dev machines.

**Comment by Jack-023** [comment]
I am getting the same popup when running the formatter this way. The first line of output from your posted logs `setting an array with 0 elements as diag for URI: ` seems to be produced whenever the editor window becomes focused, with the path being the file open in the editor.

When I try and run the formatter via `Format Document With...` the output shows that it is actually running gdformat but with errors. Interestingly, I don't get the same errors when running from the command line. I'm guessing that the extension is using a different version of gdtoolkit?

Also when I set the extension as the default formatter for GDScript and try to run `Format Document` or let it trigger on save, I instead get no feedback from the editor, with the output window displaying 
```
stderr: 
err: null
```

I only just installed this extension this morning, and it worked briefly but then these issues started happening and reinstalling has not helped. 

For additional context I am running:
VSCode						1.90.1
GDScript Formatter & Linter 		1.3.1
CLI version of gdtoolkit			4.2.2

and while I don't think it is conflicting in any way, I am also running the extension `geequlim.godot-tools`

Windows 11

Hope that helps.

**Comment by kaiharuto** [comment]
> I am getting the same popup when running the formatter this way. The first line of output from your posted logs `setting an array with 0 elements as diag for URI: ` seems to be produced whenever the editor window becomes focused, with the path being the file open in the editor.
> 
> When I try and run the formatter via `Format Document With...` the output shows that it is actually running gdformat but with errors. Interestingly, I don't get the same errors when running from the command line. I'm guessing that the extension is using a different version of gdtoolkit?
> 
> Also when I set the extension as the default formatter for GDScript and try to run `Format Document` or let it trigger on save, I instead get no feedback from the editor, with the output window displaying
> 
> ```
> stderr: 
> err: null
> ```
> 
> I only just installed this extension this morning, and it worked briefly but then these issues started happening and reinstalling has not helped.
> 
> For additional context I am running: VSCode 1.90.1 GDScript Formatter & Linter 1.3.1 CLI version of gdtoolkit 4.2.2
> 
> and while I don't think it is conflicting in any way, I am also running the extension `geequlim.godot-tools`
> 
> Windows 11
> 
> Hope that helps.

This pretty much sums up my experience, except that I'm on PopOS 22.04 LTS

**Comment by Jack-023** [comment]
>The extension uses the same version of gdtoolkit that you have installed and doesn't come with its own.

So the differences in output I was seeing with the CLI were actually because it was a different version. I didn't mention this in my previous post but when I was running it via the CLI, I was using WSL while VSCode was launched through windows so the extension was using an older version that I didn't even know I had installed. (I think it may have been installed by a formatting extension I was using previously which has since been removed from the marketplace).

Just wanted to clarify that that particular issue I was seeing is unrelated. 😅

**Comment by Jack-023** [comment]
After updating the version of `gdtoolkit` that was installed in windows to 4.2.2 it looks like the extension is working as expected again. Both in-editor linter warnings as well as format on save and `Format With...`.

**Comment by EddieDover** [comment]
> After updating the version of `gdtoolkit` that was installed in windows to 4.2.2 it looks like the extension is working as expected again. Both in-editor linter warnings as well as format on save and `Format With...`.

Are you 100% sure? The issues I described below, and the popup, are still happening to me with the example script I also provided.  I even uninstalled/reinstalled gdtoolkit to make sure, but at the time of my post I was using 4.2.2

**Comment by Jack-023** [comment]
It seems there may be a few different issues at play here. The problem was definitely fixed within my project. I just created a new project and copied the example you shared and I am getting the same error. When I fix the indentation, the linting and formatting starts working again.

I tried introducing some other simple formatting issues like adding trailing spaces to lines and the formatter runs correctly and removes them, both on save and when running  `Format With...`.

I am guessing that there are some kinds of formatting errors that break the extension while others work fine. It may be the case that the issues that the previous version of `gdtoolkit` I had installed were picking up (which are not considered issues in the current version) were one example of those that break the extension, which is why updating fixed the issue for me. 

The issue that @kaiharuto has may be different altogether since it looks like the extension isn't able to find `gdtoolkit` on their system but it just surfaces the same way in the UI.

**Comment by EddieDover** [comment]
@Jack-023  Yep, I came to the same conclusion.  I've modified the gdlint code to properly detect these types of errors. I'm taking a liberty here and assuming that, for now, it's only thrown during indention issues, so I'm incrementing the line it's pointing at because it's the following line that's an actual problem.   I'm not sure if this is a great idea, but I'm also not sure I want to willingly display a lint error on a line that doesn't contain it.

It's obvious this extension needs unit and regression tests, so I'll get on that after we settle this issue.

For now, I can modify the gdformat portion of the code to throw the error it's actually seeing, so when that popup does happen, it will look like this:
![image](https://github.com/EddieDover/vscode-formatter-godot/assets/464286/a5537c1e-42ec-4f73-80b7-6d15000845ed)

What do you think about this, limited to 10 lines of text?  If you click ok, the process continues to fail and there are no damaging effects on the file.

You're also correct about @kaiharuto. @kaiharuto , if you see this, please let me know if you're still having an issue so I can assist you.

**Comment by EddieDover** [comment]
Also, the tookit definitely checks for those syntax problems and if found, doesn't continue on, as evidenced here:
![Code_Hu8Fz71pZ4](https://github.com/EddieDover/vscode-formatter-godot/assets/464286/2ddd84ab-1a21-454a-bea4-2a7b2fd6caa6)

**Comment by Jack-023** [comment]
I think that pop-up looks pretty good as a quick fix. It is definitely a lot clearer than "An unknown error has occurred." 😄

The behaviour I see at the moment still gives the error inline (red squiggle underneath the second line, unexpected token in hover text) so it isn't too bad but the way it highlights the whole line in your example looks pretty nice.

I wonder why the formatter doesn't fix the issue here. I guess you could argue it is more of a syntax error rather than a formatting problem? I wonder how popular python formatters handle this sort of thing. Personally I would have expected it to automatically remove the extra tab.

It might be worth raising the issue upstream and see what they say because this feels like a bug in gdtoolkit to me.

**Comment by EddieDover** [comment]
> It might be worth raising the issue upstream and see what they say because this feels like a bug in gdtoolkit to me.

 I agree. This is either a bug or a design decision. I'll have to do some research tomorrow.

**Comment by EddieDover** [comment]
We now have unit testing!

![image](https://github.com/EddieDover/vscode-formatter-godot/assets/464286/d9b6a001-f02a-44ce-9e26-2132f4cd8a86)

![image](https://github.com/EddieDover/vscode-formatter-godot/assets/464286/2df5b57e-5a79-4053-aace-1543e34013c7)

**Comment by EddieDover** [comment]
If either of you would like to test out a preview build, please contact me on Discord at `eddiedover` or email me.

**Comment by EddieDover** [comment]
@Jack-023 If you can, reach out. I've got a new build of the plugin I'd love for you to test. I actually think it may be a bit faster and more stable than the current.

**Comment by EddieDover** [comment]
This was fixed in #13
